Feature/hangfire

the-man-w-laughs opened the pull request:
Hangfire is used to send emails (background jobs, Identity microservice) and to send daily info about pending project and tasks (recurring jobs, TeamTaskHub microservice).

the-man-w-laughs committed:
Add codebase for hangfire implementation

Major changes:
- add passwords.json file to store email credentials;
- add EmailConfirmationHelper;
- add ConfigurationSectionExtensions;

the-man-w-laughs committed:
Refactor Identity

the-man-w-laughs committed:
Add hangfire to Identity

Description:
Hangfire is used to send emails uting fire and forget technique.

Major changes:
- add hangfire and dashboard configuration;
- use BackgroundJob.Enqueue for user creation;
- refactor ConfirmationEmailSender;
- add where statement to GetAllUsersAsync to return only users that have confirmed their emails;
- add HangfireCredentials class.

the-man-w-laughs committed:
Add base for hangfire implementation in TeamTaskHub

the-man-w-laughs committed:
Merge branch 'main' into feature/hangfire

the-man-w-laughs committed:
Finish demo version of hangfire implementation in TeamHub microservice

Major changes:
- add TeamMemberResponseDto and corresponding mappers;
- add reccuring job registration;
- add service registration;
- add service implementation;
- add hangfire registration.

the-man-w-laughs committed:
Make automapper profiles more specific.

Major changes:
- add profile for comment.

the-man-w-laughs committed:
Refactor hangfire implementation.

Major changes:
- rename classes;
- create CustomSmtpClient to incapsulate configuration logic;
- remove redundant EntityTypeConfiguration.

the-man-w-laughs committed:
Refactor automapper profiles

Major changes:
- delete unused Dto;
- add specific fields mappers.

mebry commented on src/services/Identity/Identity.Application/Ports/Services/IEmailService.cs:
```diff
@@ -0,0 +1,7 @@
+{
+    public interface IEmailConfirmationService
+    {
+        Task ConfirmEmail(string token, string id);
```
For asynchronous methods, it's necessary to append Async to the end of the method name

mebry commented on src/services/Identity/Identity.Application/Ports/Utils/IConfirmationEmailSender.cs:
```diff
@@ -0,0 +1,9 @@
+{
+    public interface IConfirmationEmailSender
+    {
+        public Task SendEmail(AppUser appUser);
```
For asynchronous methods, it's necessary to append Async to the end of the method name

mebry commented on src/services/Identity/Identity.Application/Utils/ConfirmationEmailSender.cs:
```diff
@@ -0,0 +1,87 @@
+            var targetEmail = appUser.Email!;
+            var confirmationLink = BuildConfirmationLink(targetEmail, token);
+
+            string Body = CreateBody(confirmationLink);
```
Use var instead of explicitly specifying the type

mebry commented on src/services/Identity/Identity.Application/Utils/ConfirmationEmailSender.cs:
```diff
@@ -0,0 +1,87 @@
+                Query =
+                    $"token={Uri.EscapeDataString(token)}&email={Uri.EscapeDataString(userEmail)}"
+            };
+
```
Use options pattern

mebry commented on src/services/Identity/Identity.Application/Utils/ConfirmationEmailSender.cs:
```diff
@@ -0,0 +1,87 @@
+                    _options.Email,
+                    _options.AppPassword
+                )
+            )
```
The ConfirmationEmailSender class is tied to a specific implementation of CustomSmtpClient, making it less adaptable to changes and testing. It is recommended to use an abstraction

mebry commented on src/services/Identity/Identity.Application/Utils/ConfirmationEmailSender.cs:
```diff
@@ -0,0 +1,87 @@
+
+            string Body = CreateBody(confirmationLink);
+
+            var mailMessage = CreateMailMessage(targetEmail, Body);
```
Consider the possibility of extracting these methods into a separate service

mebry commented on src/services/Identity/Identity.WebAPI/Controllers/EmailController.cs:
```diff
@@ -0,0 +1,29 @@
+        _emailService = emailService;
+    }
+
+    [HttpGet("confirm-email")]
```
Invalid route naming

mebry commented on src/services/Identity/Identity.WebAPI/Controllers/EmailController.cs:
```diff
@@ -0,0 +1,29 @@
+
+        return Ok(
+            "Congratulations! Your email has been successfully confirmed. Thank you for verifying your email address!"
+        );
```
Create a separate file to store constants

mebry commented on src/services/Identity/Identity.WebAPI/Controllers/EmailController.cs:
```diff
@@ -0,0 +1,29 @@
+    }
+
+    [HttpGet("confirm-email")]
+    public async Task<IActionResult> ConfirmEmail(
```
It's necessary to append Async to the end of the method name

mebry commented on src/services/TeamHub/TeamHub.BLL/Contracts/IMailingService.cs:
```diff
@@ -0,0 +1,7 @@
+{
+    public interface IMailingService
+    {
+        public Task SendPendingTasks();
```
Add Async to the end of the method name

mebry commented on src/services/TeamHub/TeamHub.BLL/Contracts/IScheduledEmailService.cs:
```diff
@@ -0,0 +1,7 @@
+{
+    public interface IScheduledEmailService
+    {
+        Task Schedule();
```
The same

mebry commented on src/services/TeamHub/TeamHub.BLL/Services/MailingService.cs:
```diff
@@ -0,0 +1,123 @@
+
+                    var body = ComposeMailBody(user, teamTeamberResponseDtos);
+                    var massage = CreateMailMessage(user.Email, body);
+
+                    try
+                    {
+                        client.Send(massage);
+                    }
+                    catch (Exception ex) { }
```
Why are you using a try-catch block here?

the-man-w-laughs replied:
in case of something went wrong service continues send emails to other clients

mebry commented on src/services/TeamHub/TeamHub.BLL/Services/HolidayService.cs:
```diff
@@ -0,0 +1,43 @@
+                new DateTime(DateTime.Now.Year, 11, 1), // All Saints' Day
+                new DateTime(DateTime.Now.Year, 11, 11), // Armistice Day
+                new DateTime(DateTime.Now.Year, 12, 25) // Christmas Day
+            };
```
I can't say this is a good approach. Consider the possibility of moving this information to an external data source

mebry commented on src/services/TeamHub/TeamHub.BLL/Services/MailingService.cs:
```diff
@@ -0,0 +1,123 @@
+                        user.TeamMembers
+                    );
+
+                    var body = ComposeMailBody(user, teamTeamberResponseDtos);
+                    var massage = CreateMailMessage(user.Email, body);
```
Consider the possibility of extracting these methods into a separate service

the-man-w-laughs committed:
Alter recurring job options to correspond UTC+03:00.

the-man-w-laughs committed:
Resolve mentor's comments on hangfire implementation

Major changes:
- add async suffix where appropriate;
- implement options pattern to create a email confirmation link;
- add smtp client factory;
- add ReutingOptionsExtensions to make all letters lovercase;
- fix invalid route naming;
- remove try-catch block where client.Send(message) is invoked (even if email doesnt exists no exception is thrown).

the-man-w-laughs committed:
Resolve mentor's comments on hangfire implementation

Major changes:
- create IDailyMailMessageBuilder and IMailMessageBuilder;
- create Constants class.

the-man-w-laughs committed:
Make the holidays stored in the database

Major changes:
- add seed configuration;
- add holiday model;
- add model configuration;
- add repository.

the-man-w-laughs committed:
Configure kebab-case in each microservice.

mebry commented on src/Shared/Extensions/ConfigurationSectionExtensions.cs:
```diff
@@ -0,0 +1,20 @@
+    {
+        var name = sectionName ?? typeof(T).Name;
+        var section = configuration.GetSection(name);
+        services.Configure<T>(section);
+        return services;
```
Add an empty line before the return

mebry commented on src/services/Identity/Identity.Application/Utils/MailMessageBuilder.cs:
```diff
@@ -0,0 +1,39 @@
+            string confirmaitonLink
+        )
+        {
+            var body = CreateBody(confirmaitonLink);
+            return new MailMessage
```
Add an empty line before the return

mebry commented on src/services/TeamHub/TeamHub.BLL/Services/HolidayService.cs:
```diff
@@ -0,0 +1,32 @@
+        private async Task<bool> IsHolidayAsync(DateTime date)
+        {
+            var holidays = await _holidayRepository.GetAllAsync();
+            return holidays.Any(holiday => holiday.Date.DayOfYear == date.DayOfYear);
+        }
```
Add an empty line before the return

the-man-w-laughs committed:
Refactor code

Major changes:
- add empty line before ruturn;
- add async to asynchronous methods.
